ai_economist/foundation/components/ubi.py

```diff
@@ -0,0 +1,246 @@
+class UBI(BaseComponent):
+    """Gives each mobile agent a constant amount of coin.
+
+    Args:
+        amt_min (float): Minimum UBI payment amount. Must be >= 0 (default).
+        amt_max (float): Maximum UBI payment amount. Default is 50 (arbitrary).
+        amt_disc (float): The interval separating discrete UBI payment amounts
+            that the planner can select. Default is 5 (arbitrary).
+        period (int): Length of a period in environment timesteps. UBI is
+            distributed at the end of each period. Must be > 0. Default is 100.
+    """
+
+    name = "UBI"
+    required_entities = ["Coin"]
+    agent_subclasses = ["BasicMobileAgent", "BasicPlanner"]
+
+    def __init__(
+        self,
+        *base_component_args,
+        amt_min=0,
+        amt_max=50,
+        amt_disc=5,
+        period=100,
+        **base_component_kwargs
+    ):
+        super().__init__(*base_component_args, **base_component_kwargs)
+
+        # Min and max UBI payment amounts
+        self.amt_min = amt_min
+        assert self.amt_min >= 0
+        self.amt_max = amt_max
+
+        # Size of each discrete step
+        self.amt_disc = amt_disc
+
+        # Discrete options for the amount of each UBI payment
+        self.disc_amts = np.arange(
+            self.amt_min, self.amt_max + self.amt_disc, self.amt_disc
+        )
+        self.disc_amts = self.disc_amts[self.disc_amts <= self.amt_max]
+        assert len(self.disc_amts) > 1
+
+        # Number of discrete options for the amount of each UBI payment
+        self.n_disc_amts = len(self.disc_amts)
+
+        # Current discrete option for the size of each UBI payment
+        self.curr_amt_index = 0
+
+        # Amount of coin to give to each agent
+        self.basic_income_level = self.disc_amts[self.curr_amt_index]
+
+        # Record of the amount of UBI payment in the previous period
+        self.last_basic_income_level = None
+
+        # Record of UBI payments at each time step (for dense log)
+        self.ubi_payments = []
+
+        # How many timesteps a "year" lasts (period in which UBI is paid once)
+        self.period = int(period)
+        assert self.period > 0
+
+        # Track the UBI payment cycle
+        self.ubi_cycle_pos = 1
+
+        # Cache planner masks
+        self._planner_masks = None
+
+    def set_new_period_amt_model(self):
+        """Update UBI payment amount using actions from the planner agent."""
+
+        # AI version
+        planner_action = self.world.planner.get_component_action(
+            self.name, "UBI_payment_amount"
+        )
+        # When NO-OP, set UBI as 0 (cannot do nothing because agent may
+        # not be able to afford the same UBI rate as last time)
+        if planner_action == 0:
+            self.curr_amt_index = 0
+        elif planner_action <= self.n_disc_amts:
+            self.curr_amt_index = int(planner_action - 1)
+        else:
+            raise ValueError
+
+    def pay_ubi(self):
```

Comment:
you should check whether the planner has enough money in its inventory to pay for the UBI, so check that the planner's Coin doesn't go below 0.